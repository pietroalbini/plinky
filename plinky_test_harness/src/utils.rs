use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
