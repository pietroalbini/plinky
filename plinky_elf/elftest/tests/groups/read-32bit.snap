---
source: plinky_test_harness/src/utils.rs
assertion_line: 52
---
reading exited with exit status: 0

=== stdout ===
section .group#1 (address: 0x0)
 │
 │  group | signature: first_group#2
 │
 │    Sections:
 │  ╭──────────────────╮
 │  │ .rodata.group1#7 │
 │  ╰──────────────────╯
 ┴

section .group#2 (address: 0x0)
 │
 │  group | COMDAT | signature: second_group#3
 │
 │    Sections:
 │  ╭────────────────────╮
 │  │ .rodata.group2_1#8 │
 │  ├────────────────────┤
 │  │ .rodata.group2_2#9 │
 │  ╰────────────────────╯
 ┴

section .group#3 (address: 0x0)
 │
 │  group | signature: third_group#4
 │
 │    Sections:
 │  ╭───────────────────╮
 │  │ .rodata.group3#10 │
 │  ╰───────────────────╯
 ┴

section .rodata.group1#7 (address: 0x0, part of a group)
 │
 │  program data | permissions: R
 │
 │  ╭──────────────────────────────────┬─────────────╮
 │  │ 46 69 72 73 74 20 67 72 6f 75 70 │ First group │
 │  ╰──────────────────────────────────┴─────────────╯
 ┴

section .rodata.group2_1#8 (address: 0x0, part of a group)
 │
 │  program data | permissions: R
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ 46 69 72 73 74 20 70 61 72 74 20 6f 66 20 73 65 │ First part of se │
 │  │ 63 6f 6e 64 20 67 72 6f 75 70                   │ cond group       │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .rodata.group2_2#9 (address: 0x0, part of a group)
 │
 │  program data | permissions: R
 │
 │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │  │ 53 65 63 6f 6e 64 20 70 61 72 74 20 6f 66 20 73 │ Second part of s │
 │  │ 65 63 6f 6e 64 20 67 72 6f 75 70                │ econd group      │
 │  ╰─────────────────────────────────────────────────┴──────────────────╯
 ┴

section .rodata.group3#10 (address: 0x0, part of a group)
 │
 │  program data | permissions: R
 │
 │  ╭──────────────────────────────────┬─────────────╮
 │  │ 54 68 69 72 64 20 67 72 6f 75 70 │ Third group │
 │  ╰──────────────────────────────────┴─────────────╯
 ┴

section .symtab#11 (address: 0x0)
 │
 │    Symbol table:
 │  ╭────────────────┬─────────┬──────┬────────────┬───────┬──────╮
 │  │ Name           │ Binding │ Type │ Definition │ Value │ Size │
 │  ├────────────────┼─────────┼──────┼────────────┼───────┼──────┤
 │  │ #0             │ Local   │ -    │ Undefined  │ 0x0   │ 0x0  │
 │  ├────────────────┼─────────┼──────┼────────────┼───────┼──────┤
 │  │ groups.S#1     │ Local   │ File │ Absolute   │ 0x0   │ 0x0  │
 │  ├────────────────┼─────────┼──────┼────────────┼───────┼──────┤
 │  │ first_group#2  │ Local   │ -    │ .group#1   │ 0x0   │ 0x0  │
 │  ├────────────────┼─────────┼──────┼────────────┼───────┼──────┤
 │  │ second_group#3 │ Local   │ -    │ .group#2   │ 0x0   │ 0x0  │
 │  ├────────────────┼─────────┼──────┼────────────┼───────┼──────┤
 │  │ third_group#4  │ Local   │ -    │ .group#3   │ 0x0   │ 0x0  │
 │  ╰────────────────┴─────────┴──────┴────────────┴───────┴──────╯
 ┴


no stderr present

